### PR TITLE
[Bug fix] div: don't let the fp32 residual refinement destroy a/inf

### LINKED ISSUE #52114: ttnn.div(a, inf) returns NaN instead of 0 for finite a (fp32)
Re-filing manually. The original submission (#52112) was posted through tooling in error, which conflicts with the AI Tool Restrictions in CONTRIBUTING.md; I've closed it with a comment of apology there. The content and analysis below are unchanged through my submission

## Summary

With fp32 DEST accumulation enabled, `ttnn.div(a, inf)` for any finite `a` returns `NaN`.
IEEE-754 requires `0`. The multiply that produces the quotient gets the right answer; the
Markstein residual refinement that runs afterwards destroys it.

The registered golden for this op is `torch.div` (`ttnn/ttnn/operations/binary.py#L321-L327`),
and `torch.div(torch.tensor(1.0), torch.tensor(float("inf")))` is `0.0`, so the expected value
is not in dispute.

## Where

`calculate_sfpu_binary_div`, in the `is_fp32_dest_acc_en` branch:

- blackhole: `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h`
- wormhole_b0: `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h`

```cpp
sfpi::vFloat r = sfpu_reciprocal_iter<2>(in1);
sfpi::vFloat result = in0 * r;
if constexpr (is_fp32_dest_acc_en) {
    v_if(sfpi::exexp(result, sfpi::ExponentMode::Biased) != 255) {
        sfpi::vFloat e = in0 - result * in1;
        result = result + e * r;
    }
    v_endif;
}
```

## Trace

For finite `a` and `b = +inf`:

| step | value |
|---|---|
| `r = sfpu_reciprocal_iter<2>(inf)` | `+0` — documented in `ckernel_sfpu_recip.h` |
| `result = a * 0` | `0` — **this is already the correct answer** |
| guard `exexp(result) != 255` | biased exponent of `0` is `0`, so the guard is **true** and refinement runs |
| `e = a - result*b = a - (0 * inf)` | `0 * inf` is `NaN`, so `e = NaN` |
| `result = 0 + NaN * 0` | **`NaN`** |

The existing guard only skips refinement when the *quotient* is already non-finite — the
`a/b → inf` overflow case. It does not consider a non-finite *denominator*, which is the case
where the residual `result * den` is an invalid `0 * inf` operation.

The trailing special-case block handles `in1 == 0` (`x/0 → ±inf`, `0/0 → NaN`). There is no
corresponding handling for `in1 == ±inf`.

## Scope

- Affects the `is_fp32_dest_acc_en` path of `calculate_sfpu_binary_div` only.
- The plain `BinaryOp::DIV` case in `calculate_sfpu_binary` has no refinement, so `a/inf` is
  `a * 0 = 0` there and appears correct.
- Quasar has no Markstein block in its div path and appears unaffected.
- `div_no_nan` shares the helper and should be checked alongside.

## Why the tests miss it

`test_binary_fp32.py::test_div_fp32` includes an `inf / inf` pair, which correctly yields `NaN`
and passes. It does not test **finite / inf**, which is the failing combination. The sweeps draw
both operands from `[-100, 100]` and never generate infinities at all.

## Suggested fix

Extend the refinement guard to also skip when the denominator is non-finite, e.g.
`v_if(exexp(result, Biased) != 255 && exexp(in1, Biased) != 255)`, or equivalently skip when
`r == 0`. Either keeps `result = a * 0 = 0` and leaves every finite case untouched.

## Caveat on verification

I do not have hardware, so this is derived from the source plus a float32 model of the code path
rather than measured on device. It assumes the SFPU multiply follows IEEE for `0 * inf → NaN`.
That assumption is consistent with this file's own comments, which already reason about
`inf + (-inf) = NaN` in the guard directly above. If the SFPU instead defines `0 * inf` as `0`,
this does not reproduce and the issue can be closed — a one-line check on device would settle it
either way.

## Why it matters

Like #52094, this is not precision loss: it is a wrong, non-finite answer where the correct
answer is an ordinary `0`. Unlike a silent `0`, a `NaN` does propagate loudly — but it appears in
exactly the code that is trying to be *more* accurate, and it turns a case the hardware already
got right into a failure. Dividing by an accumulated magnitude that has overflowed to `inf` is a
realistic way to hit it.

### PR DESCRIPTION
**Pull request by kingsleytorlowei**
### PR Category

Bug fix

### Summary

Closes #52114.

With fp32 DEST accumulation enabled, `ttnn.div(a, inf)` returns `NaN` for every finite `a`. IEEE-754 and the registered `torch.div` golden both give a signed zero.

There is no divide instruction, so `calculate_sfpu_binary_div` evaluates `a / b` as `a * (1/b)`. For `b = ±inf` the reciprocal is `±0`, so the product is already the answer. The fp32 residual (Markstein) refinement then runs anyway and destroys it:

```
r      = 1/inf         = +0
result = a * 0         = 0      <- already correct
guard  : exponent of 0 is 0, not 255, so the guard passes
e      = a - (0 * inf) = NaN
result = 0 + NaN * 0   = NaN
```

The guard tests whether the *quotient* is non-finite, which is the right check for `a / tiny -> inf`. It does not consider a non-finite *denominator*, where the quotient is an ordinary zero and the gate opens. So a precision step overwrites a result the hardware had already computed correctly.

This extends the guard to the denominator, and sets the NaN case explicitly for the reason below. Applied to both `blackhole` and `wormhole_b0`.

### Notes for reviewers

**Widening the guard alone is not sufficient.** The issue suggests skipping refinement whenever the denominator is non-finite. I tried that first. A biased exponent of 255 tags NaN as well as infinity, so it regresses `a / NaN`:

| case | torch | before | guard widened alone | this PR |
|---|---|---|---|---|
| `a / inf` | `0` | `NaN` | `0` | `0` |
| `a / NaN` | `NaN` | `NaN` | `0` | `NaN` |

`sfpu_reciprocal_iter` returns `0` rather than `NaN` for a NaN input, so `result = a * 0 = 0`, and it was the residual `e = a - 0 * NaN = NaN` that restored the NaN. NaN division was only correct as a side effect of the same step that breaks `a / inf`. Setting it explicitly also removes that dependency on the reciprocal's behaviour. The issue's other suggestion, skipping when `r == 0`, is not equivalent either: sfpi's `!=` does not treat `-0` and `+0` the way IEEE
does, so it misfires on `-inf`.

**Why two conditionals rather than one `!is_inf(in1)`.** That is the natural spelling and was my first attempt. It crashes the pinned SFPI 7.74.0 toolchain when the div kernel is JIT-compiled:

```
during GIMPLE pass: rvtt_expand
internal compiler error: Segmentation fault
0x9fbdab update_ssa(unsigned int)
gcc (tenstorrent/sfpi:7.74.0[897]) 15.1.0
```

Every three-predicate form reproduced it, including
`exexp(in1) != 255 || exman(in1) != 0`. Two predicates compile, and unmodified `main` compiles fine on the same toolchain. Happy to file that against tenstorrent/sfpi, and equally happy to switch if someone knows a spelling that
survives.

**Testing.** Blackhole simulator, cold kernel cache per run.

| check | result |
|---|---|
| new test on the parent commit | fails, `Tensors are not finite at the same positions` |
| new test against the widened-guard version | fails on `a / NaN` |
| new test with this change | passes |
| `test_binary_fp32.py` + `test_binaryng_fp32.py` | 107 passed, 1 skipped |

The NaN pairs in the test are the regression guard for the first note above, not decoration. `test_div_fp32` previously covered only `inf / inf`, which passes even when broken because that quotient is already non-finite, and the sweeps
elsewhere in the file draw from `[-100, 100]` and never generate non-finite values.

**Limits I'd rather state than have found.** Simulator, not silicon. The
`wormhole_b0` copy was never executed, since that simulator rejects these fp32 tests for an unrelated reason (`tensix_unpacr: unpack_to_dst=0 in_data_format=0 out_data_format=0`) including tests this change does not touch. And the *sign* of the zero is reasoned rather than measured, because `assert_with_ulp` treats `-0`
and `+0` as equal and the simulator does not preserve the sign of zero through an fp32 multiply.

**Found nearby, deliberately not fixed here.** `a / NaN` also returns `0` on the bfloat16 path, from the same `reciprocal(NaN) -> 0` behaviour; that path has no refinement so it never had the accidental rescue. And `1.0 / 1e38` returns `0` where torch gives `1e-38`, which is the separate underflow issue #52094. Happy to file the first.

### CODE REVIEW

**Comment by kingsleytorlowei** [comment]
Closing in favour of #53768, which fixes the same issue more completely. It also corrects the sign of the zero, fixes `-0.0 / 0.0` returning `-inf` instead of NaN, and enrols `SfpuElwdiv` in the tt-llk special-value sweeps, which is the reason none of this was caught in the first place.

I've moved the one finding that isn't covered there, a toolchain crash, into a comment on that PR.
